### ISSUE TITLE
[iOS] Using sd_setImage doesn't let the ViewController do deinit properly 

### ISSUE THREAD

**Issue by jenjia** [rs/SDWebImage#2205]
### New Issue Checklist

* [x ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x ] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios 
 Platform Version        | 11.0 
 SDWebImage Version      |  4.2.0
 Integration Method      |  cocoapods 
 Xcode Version           | Xcode 9 
 Repro rate              | all the time (100%) 
 Repro with our demo prj | N/A
 Demo project link       | N/A

### Issue Description and Steps
In my ViewController I set the function deinit {} to properly know when the view is deleted, as it follows:
```
deinit {
        print("deinit.\(self)")
    }
```
I want to set a image using SDWebImage into a UIImageView, named imageView, using the following code:
```
self.imageView.sd_setImage(
with: url, 
placeholderImage: #imageLiteral(resourceName: "loadingImg"), 
options: [SDWebImageOptions.progressiveDownload], 
completed: { (image, err, cacheType, url) in
     print("SDWebImage...... 1")
            if err != nil {
                 print("SDWebImage...... 2")
                  self.imageView.image = #imageLiteral(resourceName: "noImg")
                  return
             }else{
                  print("SDWebImage...... 3")
                   return
               }
})
```
If I use this this way, I get the correct output:
```
SDWebImage...... 1
SDWebImage...... 3
```
but, when I click "Back" to go to the previous ViewController this one is not correctly deinitialized.
On the other hand, if I just comment the setting of the "noImg", as shown:
```
self.imageView.sd_setImage(
with: url, 
placeholderImage: #imageLiteral(resourceName: "loadingImg"), 
options: [SDWebImageOptions.progressiveDownload], 
completed: { (image, err, cacheType, url) in
                    print("SDWebImage...... 1")
                    if err != nil {
                        print("SDWebImage...... 2")
                        //self.imageView.image = #imageLiteral(resourceName: "noImg")
                        return
                    }else{
                        print("SDWebImage...... 3")
                        return
                    }

                })
```
This way, if I click the "Back" button, the output is the following:
```
SDWebImage...... 1
SDWebImage...... 3
deinit.<XXXX.ViewController: 0x10455ed20>
```
which is what should always happen.

Am I missing something?
Thank you.

PS: In the meantime I am using the following code (except this way I lose the progressive download, and don't know if is the best way to do it ):
```
self.imageView.image = #imageLiteral(resourceName: "loadingImg")
SDWebImageManager.shared().imageDownloader?.downloadImage(
with: url, 
options: [], 
progress: nil, 
completed: { (image, data, error, success) in
                    if image != nil {
                        self.imageView.image = image
                    }else{
                        self.imageView.image = #imageLiteral(resourceName: "noImg")
                    }
                })
```

**Comment by dreampiggy** [comment]
The completion block is captured into `SDWebImageManager` and will be only removed after the load progress finished (Cache fetched or network finished). So if you use any `self` in the completionBlock, the dealloc may be be later than you expected. (Because only after network finished, the strong reference to completionBlock is removed and your VC dealloc)

If you do not want this, just mark it to weak self and strong self. For Objective-C, use this:

```objective-c
__weak typeof(self) wself = self;
[self.imageView sd_setImageWithURL:url completed:^(UIImage * _Nullable image, NSError * _Nullable error, SDImageCacheType cacheType, NSURL * _Nullable imageURL) {
    __strong typeof(wself) sself = wself;
    sself.imageView.image = image;
}];
```

Or use the `unowned` in Swift:

```swift
self.imageView.sd_setImage(with: url, completed: { [unowned self] (image, error, cacheType, url) in
    self.imageView.image = image
})
```

Actually, I think this question should add to [common problems](https://github.com/rs/SDWebImage/wiki/Common-Problems) :)

**Comment by jenjia** [comment]
Adding the `[weak self]` worked perfectly. 
Never known about this trick.
We are always learning. Sorry to bother. 

You also need to add `?` after in the `self`, as shown:
`self?.imageView.image = image`

I also advice to add that to the common problems, I guess a lot of people will/is missing this.

Nevertheless, in the example I provided, the closure just finished correctly. The image is shown correctly. Sometimes it passed a bit of time before I go back. 
That said, I am trying to figure out why it is really happening, that it takes so long to release the `self`. Mere curiosity.
Cheers

**Comment by dreampiggy** [comment]
Added [Handler self capture in completion block](https://github.com/rs/SDWebImage/wiki/Common-Problems#handler-self-capture-in-completion-block) in the common problems.

This is not a issue or bug for SDWebImage because it's about Objective-C/Swift 's memory management. Some iOS developer may be not familiar with this. We should provide the correct way to do this :)

**Comment by bpoplauschi** [comment]
@dreampiggy 👍 for creating the Common problems doc
